src/Microsoft.Extensions.Configuration.AzureAppConfiguration/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,186 @@
+        }
+
+        public AsyncPageable<ConfigurationSetting> GetConfigurationSettingsAsync(SettingSelector selector, CancellationToken cancellationToken = default)
+        {
+            return ExecuteWithFailOverPolicy(clients.Select((client) => new Func<AsyncPageable<ConfigurationSetting>>(() => client.Client.GetConfigurationSettingsAsync(selector, cancellationToken))), cancellationToken);
+        }
+
+        public void UpdateSyncToken(Uri endpoint, string syncToken)
+        {
+            this.clients.Single(client => client.Endpoint.Host.Equals(endpoint.Host)).Client.UpdateSyncToken(syncToken);
+        }
+
+        private async Task<Response<T>> ExecuteWithFailOverPolicyAsync<T>(IEnumerable<Func<Task<Response<T>>>> delegates, CancellationToken cancellationToken = default)
+        {
+            Exception latestException = null;
+
+            for (var i = ShouldTryPrimaryConfigStore() ? 0 : 1; i < clients.Count(); i++)
+            {
+                var success = false;
+                var operationCanceled = false;
+
+                try
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    var result = await delegates.ElementAt(i)().ConfigureAwait(false);
+                    success = true;
+
+                    return result;
+                }
+                catch (RequestFailedException e) when (e.Status == (int)HttpStatusCode.ServiceUnavailable || e.Status == HttpStatusRequestThrottled)
+                {
+                    latestException = e;
+                    continue;
+                }
+                catch (AggregateException ex) when (ex?.InnerExceptions?.All(e => e is RequestFailedException) == true)
+                {
+                    var lastException = ex.InnerExceptions.LastOrDefault() as RequestFailedException;
+
+                    if (lastException?.Status == (int)HttpStatusCode.ServiceUnavailable || lastException?.Status == HttpStatusRequestThrottled)
+                    {
+                        latestException = lastException;
+                        continue;
+                    }
+
+                    throw;
+                }
+                catch (OperationCanceledException)
+                {
+                    operationCanceled = true;
+                    throw;
+                }
+                finally
+                {
+                    if (i == 0 && !operationCanceled)
+                    {
+                        UpdatePrimaryConfigStoreStatus(success);
+                    }
+                }
+            }
+
+            throw latestException;
+        }
+
+        private T ExecuteWithFailOverPolicy<T>(IEnumerable<Func<T>> delegates, CancellationToken cancellationToken = default)
+        {
+            Exception latestException = null;
+
+            for (var i = ShouldTryPrimaryConfigStore() ? 0 : 1; i < clients.Count(); i++)
```

Comment:
It seems like this should be factored into creating the delegates enumerable. Not the iteration over the delegates.

Reply:
I'm confused. Would you mind elaborating on this a little bit?
whether to call a delegate or not is dependent on current time and whether the primary store was successful in the last attempt.
I'm missing something here.

Reply:
It's odd for the function to take delegates and skip over them. Why not just pass the delegates that should be used. The caller should be able to construct that efficiently.

Also, now the logic has to be repeated in the two execution wrappers. What if instead callers did something like `ExecuteWithFailOverPolicyAsync(GetEligibleClients().Select((client) => new Func<Task<Response<ConfigurationSetting>>>(() =>`

Then client selection goes in one place, and I think the failover policy is cleaner.

Reply:
I suppose that breaks down a bit because this code is assuming primary client is always encapsulated by the first delegate.